ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -169,22 +196,172 @@ module Json = struct
+    let add_first = function
+      | head :: rest ->
+          let head = `O (("first", `Bool true) :: of_param head) in
+          let rest =
+            List.map
+              (fun item -> `O (("first", `Bool false) :: of_param item))
+              rest
+          in
+          head :: rest
+      | [] ->
+          []
+    in
+    params
+    |> List.map (fun p ->
+           let fp_type = func_name_suffix p.param_type in
+           let t, _e = string_of_ty_with_enums p.param_type in
+           (t, p.param_name, p.param_doc, fp_type)
+       )
+    |> fun params -> `A (add_first params)
+
+  let of_error e = `O [("name", `String e.err_name); ("doc", `String e.err_doc)]
+
+  let of_errors = function
+    | [] ->
+        `Null
+    | errors ->
+        `A (List.map of_error errors)
+
+  let add_session_info class_name method_name =
+    match (class_name, method_name) with
+    | "session", "login_with_password"
+    | "session", "slave_local_login_with_password" ->
+        [("session_login", `Bool true); ("session_logout", `Bool false)]
+    | "session", "logout" | "session", "local_logout" ->
+        [("session_login", `Bool false); ("session_logout", `Bool true)]
+    | _ ->
+        [("session_login", `Bool false); ("session_logout", `Bool false)]
+
+  let desc_of_msg msg ctor_fields =
+    let ctor =
+      if msg.msg_tag = FromObject Make then
+        Printf.sprintf " The constructor args are: %s (* = non-optional)."
+          ctor_fields
+      else
+        ""
+    in
+    match msg.msg_doc ^ ctor with
+    | "" ->
+        `Null
+    | desc ->
+        `String (String.trim desc)
+
+  let ctor_fields_of_obj obj =
+    Datamodel_utils.fields_of_obj obj
+    |> List.filter (function
+         | {qualifier= StaticRO | RW; _} ->
+             true
+         | _ ->
+             false
+         )
+    |> List.map (fun f ->
+           String.concat "_" f.full_name
+           ^ if f.default_value = None then "*" else ""
+       )
+    |> String.concat ", "
+
+  let messages_of_obj obj =
+    let ctor_fields = ctor_fields_of_obj obj in
+    let params_in_msg msg =
+      if msg.msg_session then
+        session_id :: msg.msg_params
+      else
+        msg.msg_params
+    in
+    List.map
+      (fun msg ->
+        let params = params_in_msg msg |> of_params obj.name in
+        let base_assoc_list =
+          [
+            ("method_name", `String msg.msg_name)
+          ; ("class_name", `String obj.name)
+          ; ("class_name_exported", `String (snake_to_camel obj.name))
+          ; ("method_name_exported", `String (snake_to_camel msg.msg_name))
+          ; ("description", desc_of_msg msg ctor_fields)
+          ; ("result", of_result obj msg)
+          ; ("params", params)
+          ; ("errors", of_errors msg.msg_errors)
+          ; ("has_error", `Bool (msg.msg_errors <> []))
+          ; ("async", `Bool msg.msg_async)
+          ]
+        in
+        if obj.name = "session" then
+          let header_params =
```

Comment:
I can't see the purpose of `header_params`. Would the `params` do the work?

Reply:
OK. Thanks for the explanation :)
Please add a comment for this. This is to make the `session_id` not appear in the parameter list of a function like `session.get_all_subject_identifiers` but in the serialize section.
How about:
```
      if obj.name = "session" then
        `O
          (("func_params", (msg.msg_params |> of_params))
          :: (add_session_info obj.name msg.msg_name @ base_assoc_list)
          )
      else
        `O base_assoc_list
```